File: Ruby/day09.rb
#!/usr/bin/env ruby

require 'util'
#require 'geometry'
#require 'grid'

# INPUT_FILE = 'day09_test.txt'
INPUT_FILE = 'day09_challenge.txt'

puts 'Advent of Code 2023, Day 09: Mirage Maintenance'

class Oasis
  def initialize(defn)
    @sequence = defn.split(/ +/).map { |num| Integer(num) }
    @pyramid = nil
  end
  
  def next_number
    build_pyramid if @pyramid == nil
    @pyramid[0][-1]
  end
  
  def prev_number
    build_pyramid if @pyramid == nil
    @pyramid[0][0]
  end
  
  def build_pyramid
    @pyramid = [@sequence.dup]

    i = 0
    while !all_zeroes(@pyramid[i])
      next_row = []
      for j in (0 ... @pyramid[i].length-1) do
        next_row << @pyramid[i][j+1] - @pyramid[i][j]
      end
      @pyramid << next_row
      i += 1
    end
    
    # Add the first and last values
    @pyramid[i].unshift(0)
    @pyramid[i].push(0)
    i -= 1
    while i >= 0
      @pyramid[i].unshift(@pyramid[i][0] - @pyramid[i+1][0])
      @pyramid[i].push(@pyramid[i][-1] + @pyramid[i+1][-1])
      i -= 1
    end
  end
  
  def all_zeroes(arr)
    for val in arr do
      return false if val != 0
    end
    true
  end
end

def solve_part_one(oases)
  numbers = oases.map { |oasis| oasis.next_number }
  sum = numbers.inject(:+)
  puts "Part One: the sum of next numbers is #{sum}"
end


def solve_part_two(oases)
  numbers = oases.map { |oasis| oasis.prev_number }
  sum = numbers.inject(:+)
  puts "Part Two: the sum of previous numbers is #{sum}"
end


input1 = Util.read_grouped_input("../Input/#{INPUT_FILE}", 0)
oases = input1.map { |line| Oasis.new(line) }
solve_part_one oases

# input2 = Util.read_grouped_input("../Input/#{INPUT_FILE}", 0)
# oases = input2.map { |line| Oasis.new(line) }
solve_part_two oases

File: Perl/day09.pl
#!/usr/bin/env perl
use v5.40;

our $directory;
BEGIN { use Cwd; $directory = cwd; }
use lib $directory . '/lib';

use feature 'signatures';
use Data::Printer;
use List::Util qw( all sum );

use AOC::Util;

my $DAY = sprintf("%02d", 9);
my $INPUT = $ARGV[0];
my $INPUT_FILE = 'day' . $DAY . '_' . $INPUT . '.txt';

my @input = read_grouped_input("../input/$INPUT_FILE", 0);

say "Advent of Code 2023, Day 9: Mirage Maintenance";

my @histories = map { build_history($_) } @input;

solve_part('One');
solve_part('Two');

exit( 0 );

sub solve_part($part) {
	my @extrapolated = ();
	for my $h (@histories) {
		my @h = @{$h};
		my $ext_num = 0;
		for (my $i = $#h-1; $i >= 0; $i--) {
			if ($part eq 'One') {
				$ext_num = $h[$i][-1] + $ext_num;
			}
			else {
				$ext_num = $h[$i][0] - $ext_num;
			}
		}
		push(@extrapolated, $ext_num);
	}
	my $total = sum(@extrapolated);
	say "Part $part: the total of the extrapolated values is $total.";
}

sub build_history($line) {
	my @h = ();
	my @first_row = split(/ +/, $line);
	push(@h, \@first_row);
	my @diffs;
	do {
		my @last_row = @{$h[$#h]};
		my @d = mk_diffs(@last_row);
		push(@h, \@d);
		@diffs = @d;
	} until (all { $_ == 0 } @diffs);
	return \@h;
}

sub mk_diffs(@nums) {
	my @d = ();
	for my $i (1..$#nums) {
		push( @d, $nums[$i] - $nums[$i-1]);
	}
	return @d;
}
